lib/csv/row.rb

```diff
@@ -659,8 +659,26 @@ def to_h
     end
     alias_method :to_hash, :to_h
 
+    # :call-seq:
+    #   row.deconstruct_keys(keys) -> hash
+    #
+    # Returns the new \Hash suitable for pattern matching containing only the
+    # keys specified as an argument.
+    def deconstruct_keys(keys)
+      keys.to_h { |key| [key, self[key]] }
```

Comment:
The keys, as supplied, will be `Symbol`s. This can present an issue as CSV rows will more than likely be `String` keyed:

```ruby
require 'csv'
# => true

csv = CSV.parse <<~RUBY, headers: true
  a,b,c
  1,2,3
  2,3,4
  4,5,6
RUBY
# => #<CSV::Table mode:col_or_row row_count:4>

csv.headers
# => ["a", "b", "c"]

csv.to_a
# => [["a", "b", "c"], ["1", "2", "3"], ["2", "3", "4"], ["4", "5", "6"]]

csv.first.class
# => CSV::Row

csv.first.fields
# => ["1", "2", "3"]
csv.first.headers
# => ["a", "b", "c"]
```

If you want to support this you would want to instead transform the keys to `String`, but this may be controversial as it conflates the two.

Reply:
This could be resolved one of a few ways.

In the first, retrieval of values would be done versus the `String` equivalent of the value:

```ruby
def deconstruct_keys(keys)
  keys.to_h { |k| [k, self[k.to_s]] }
end
```

This assumes, however commonly it may be, that all keys are `String` values. This could be resolved via:

```ruby
def deconstruct_keys(keys)
  first_header = headers.first
  case first_header
  when String then keys.to_h { |k| [k, self[k.to_s]] }
  when Symbol then keys.to_h { |k| [k, self[k]] }
  else {}
  end
end
```

...though this is far more verbose. I do not think that `Symbol` keys are commonly used in the context of `CSV`, so this concern might be ignored safely, but I leave that up to the core group of `CSV` to decide the risk.

Reply:
I think I'd rather leave it as is for now. You can always pass `header_converters: :symbol` to the parse function, which would make this work as expected.

I think I would be really surprised if in my matching I specified symbols and it matched against strings.